sycl-fusion/jit-compiler/lib/translation/KernelTranslation.cpp

```diff
@@ -0,0 +1,286 @@
+    // store it in the SYCLKernelInfo.
+    restoreKernelAttributes(Result.get(), Kernel);
+  }
+  return std::move(Result);
+}
+
+llvm::Expected<std::unique_ptr<llvm::Module>>
+KernelTranslator::loadLLVMKernel(llvm::LLVMContext &LLVMCtx,
+                                 SYCLKernelInfo &Kernel) {
+  auto &BinInfo = Kernel.BinaryInfo;
+  llvm::StringRef RawData(reinterpret_cast<const char *>(BinInfo.BinaryStart),
+                          BinInfo.BinarySize);
+  return llvm::parseBitcodeFile(
+      MemoryBuffer::getMemBuffer(RawData, Kernel.Name,
+                                 /* RequiresNullTermnator*/ false)
+          ->getMemBufferRef(),
+      LLVMCtx);
+}
+
+llvm::Expected<std::unique_ptr<llvm::Module>>
+KernelTranslator::loadSPIRVKernel(llvm::LLVMContext &LLVMCtx,
+                                  SYCLKernelInfo &Kernel) {
+  return SPIRVLLVMTranslator::loadSPIRVKernel(LLVMCtx, Kernel);
+}
+
+llvm::Error KernelTranslator::translateKernel(SYCLKernelInfo &Kernel,
+                                              llvm::Module &Mod,
+                                              JITContext &JITCtx,
+                                              BinaryFormat Format) {
+
+  KernelBinary *KernelBin = nullptr;
+  switch (Format) {
+  case BinaryFormat::SPIRV: {
+    llvm::Expected<KernelBinary *> BinaryOrError =
+        translateToSPIRV(Mod, JITCtx);
+    if (auto Error = BinaryOrError.takeError()) {
+      return Error;
+    }
+    KernelBin = *BinaryOrError;
+    break;
+  }
+  case BinaryFormat::PTX: {
+    llvm::Expected<KernelBinary *> BinaryOrError =
+        translateToPTX(Kernel, Mod, JITCtx);
+    if (auto Error = BinaryOrError.takeError()) {
+      return Error;
+    }
+    KernelBin = *BinaryOrError;
+    break;
+  }
+  default: {
+    return createStringError(
+        inconvertibleErrorCode(),
+        "Failed to translate kernel to unsupported output format");
+  }
+  }
+
+  // Update the KernelInfo for the fused kernel with the address and size of the
+  // SPIR-V binary resulting from translation.
+  SYCLKernelBinaryInfo &FusedBinaryInfo = Kernel.BinaryInfo;
+  FusedBinaryInfo.Format = Format;
+  // Output SPIR-V should use the same number of address bits as the input
+  // SPIR-V. SPIR-V translation requires all modules to use the same number of
+  // address bits, so it's safe to take the value from the first one.
+  FusedBinaryInfo.AddressBits = Mod.getDataLayout().getPointerSizeInBits();
+  FusedBinaryInfo.BinaryStart = KernelBin->address();
+  FusedBinaryInfo.BinarySize = KernelBin->size();
+  return Error::success();
+}
+
+llvm::Expected<KernelBinary *>
+KernelTranslator::translateToSPIRV(llvm::Module &Mod, JITContext &JITCtx) {
+  return SPIRVLLVMTranslator::translateLLVMtoSPIRV(Mod, JITCtx);
+}
+
+llvm::Expected<KernelBinary *>
+KernelTranslator::translateToPTX(SYCLKernelInfo &KernelInfo, llvm::Module &Mod,
+                                 JITContext &JITCtx) {
+#ifndef FUSION_JIT_SUPPORT_PTX
```

Comment:
NIT: Can you extend the conditional compilation guard so that the `else` region holds the whole function?